compiler/rustc_mir_build/src/build/coverageinfo.rs

```diff
@@ -199,17 +203,28 @@ impl BranchInfoBuilder {
 /// This limit may be relaxed if the [upstream change](https://github.com/llvm/llvm-project/pull/82448) is merged.
 const MAX_CONDITIONS_NUM_IN_DECISION: usize = 6;
 
-struct MCDCState {
+#[derive(Default)]
+struct MCDCDecisionCtx {
     /// To construct condition evaluation tree.
     decision_stack: VecDeque<ConditionInfo>,
     processing_decision: Option<MCDCDecisionSpan>,
 }
 
+struct MCDCState {
+    decision_ctx_stack: Vec<MCDCDecisionCtx>,
+}
+
 impl MCDCState {
     fn new_if_enabled(tcx: TyCtxt<'_>) -> Option<Self> {
         tcx.sess
             .instrument_coverage_mcdc()
-            .then(|| Self { decision_stack: VecDeque::new(), processing_decision: None })
+            .then(|| Self { decision_ctx_stack: vec![MCDCDecisionCtx::default()] })
+    }
+
+    #[inline]
+    fn decision_depth(&self) -> u16 {
+        self.decision_ctx_stack.len().checked_sub(1).expect("Unexpected empty decision stack")
+            as u16
```

Comment:
This might never matter, but having an unchecked `as u16` cast here seems wrong.

If someone does manage to exceed 2^16 nested contexts, it's probably better to loudly panic than to silently do the wrong thing.

Reply:
Actually, what's the reason for limiting decision depth to `u16` in the first place?

Is it mainly to avoid increasing the size of `CoverageKind`?

Reply:
In this design rust would allocate temp variables on the stack with number of max decision depth. Thus a function having decisions whose depth is 65535 can allocate about 256KB (65535*4) additional memory only for mcdc, which seems large enough. 
Human does not likely write decisions having hundreds of depth. These decisions probably are generated by complex macros.
This may be fixed to `u16::try_from` or even `u8::try::from`.

Reply:
Normally I would prefer to avoid an arbitrary limit without a useful reason.

But in this case, I think “avoid increasing the size of `CoverageKind`” is a good enough reason, combined with the fact that a nesting depth of 2^16 is unrealistic.

So I agree that this can just be fixed with `u16::try_from`. But we should also have some comments explaining why.

Reply:
(The alternative would be to use `usize`, and then box the contents of `CoverageKind::CondBitmapUpdate`. But I don't think that's worth the hassle.)